feedback_prompt_learning/search_algo/mcts.py

```
@@ -0,0 +1,931 @@
+# MCTS OPTIMIZER WITH FEEDBACK
+# ============================================================================
+
+class MCTSPromptOptimizerFeedback:
+    """MCTS for Prompt Optimization using LLM-generated actions based on feedback"""
+    
+    def __init__(
+        self,
+        initial_prompt: str,
+        train_dataset: List[Tuple[str, str]],
+        reward_fn: Callable[[str, str, str], Tuple[float, str]],
+        llm: ChatOpenAI = hydra_instantiate(cfg.optimizer.llm.student),
```

Ruff (0.14.7)

Do not perform function call `hydra_instantiate` in argument defaults; instead, perform the call within the function, or read the default from a module-level singleton variable

(B008)